### PR TITLE
Rewrite all material including README and examples

### PR DESCRIPTION
**Pull request by robertmcnees**
This PR is a near complete rewrite of the existing guide. 

The existing guide was dated in several ways, most notably in a reference to `@EnableDiscoveryClient` (identified in issue #25) and `boostrap.properties` (identified in issue #28). Both of these methods are not available or recommended with the latest version of Spring Cloud. Rather than simply updating the existing applications to work, I took the opportunity to rethink the examples used.

A couple of key differences in this PR to what I replaced:

- Removed the `complete` and `initial` folders: I felt that having multiple applications in a single guide felt very cluttered when they were repeated in each folder. The `complete` and `initial` format works OK for simple apps, but when multiple apps are required it is my opinion that complexity increases beyond the value. If the users want to start from scratch, they can use the provided Spring Initializr links.
- Added a third application: I felt that the existing example didn't go the final step and help the user understand how service discovery could be used in a real application. To demonstrate this, I created two services that are both registered with Eureka and communicate to one another.
- I used a single build tool: For the same reason that I removed `complete` and `initial` folders, I decided against including both gradle and maven dependencies. There are several build commands specified in the `README.adoc` and I wanted to avoid having to list both gradle and maven versions.
- Added a CI Build in GitHub Actions, similar to other spring-guides.

### CODE REVIEW

**Comment by robertmcnees** [comment]
Given the significant changes, I'm hoping to have a review from @OlgaMaciaszek.

File: README.adoc
Comment (Buzzardo) [inline README.adoc]: Add "the" before "Eureka"

File: README.adoc
Comment (Buzzardo) [inline README.adoc]: Replace "if you'd like to" with "To"

Remove "please use"

Add a comma before "as defined below"

File: README.adoc
Comment (Buzzardo) [inline README.adoc]: There should be two sets of instructions here, one for the server and one for the clients.

File: README.adoc
Comment (Buzzardo) [inline README.adoc]: Remove "will"

File: README.adoc
Comment (Buzzardo) [inline README.adoc]: Change to:

You can now start the Eureka server by running `./gradlew bootRun`.

File: README.adoc
Comment (Buzzardo) [inline README.adoc]: Remove "will"

File: README.adoc
Comment (Buzzardo) [inline README.adoc]: Add a comma after "point"

File: README.adoc
Comment (Buzzardo) [inline README.adoc]: Move command to after the command.

So replace "command `./gradlew bootRun`" with "`./gradlew bootRun` command"

The other way to do it is "command, `./gradlew bootRun`," (that is, wrapping the command in commas). This has to do with English's ordering of nouns. If you shift from the usual order, you have to make it be parenthetical, so you need commas. I run into this constantly, because people speak in this order (ignoring the commas when speaking), but it's incorrect to write in this order.

File: README.adoc
Comment (Buzzardo) [inline README.adoc]: Change "are running" to "run"

File: README.adoc
Comment (Buzzardo) [inline README.adoc:124]: Change "A" to "An"

File: README.adoc
Comment (Buzzardo) [inline README.adoc]: Add "called" after "class"

Change "that exposes" to "to expose"

Change "will" to "can"

File: README.adoc
Comment (Buzzardo) [inline README.adoc]: Add "called" after "class"

Change "that exposes" to "to expose"

File: README.adoc
Comment (Buzzardo) [inline README.adoc]: Change "only has" to "has only"

File: README.adoc
Comment (Buzzardo) [inline README.adoc]: Change "line" to "following line"

File: README.adoc
Comment (Buzzardo) [inline README.adoc]: Change "have the" to "have a"

File: README.adoc
Comment (Buzzardo) [inline README.adoc]: Change "executing" to "running"

Add "following" before "command"

File: README.adoc
Comment (Buzzardo) [inline README.adoc]: Change to:

This guide walked through the following steps (which you could do by implementing the code shown throughout this guide or by using the code in the solution repository):

File: README.adoc
Comment (Buzzardo) [inline README.adoc]: Change "command `./gradlew bootRun`" to "`./gradlew bootRun` command"

File: README.adoc
Comment (Buzzardo) [inline README.adoc]: Change "command `./gradlew bootRun`" to "`./gradlew bootRun` command"

File: README.adoc
Comment (Buzzardo) [inline README.adoc]: Change "command `./gradlew bootRun`" to "`./gradlew bootRun` command"

File: README.adoc
Comment (Buzzardo) [inline README.adoc]: Change "Observe" to "Observe that"

File: README.adoc
Comment (Buzzardo) [inline README.adoc]: Change "command `curl http://localhost:8082/serviceB/helloEureka`" to "`curl http://localhost:8082/serviceB/helloEureka` command"

File: README.adoc
Comment (Buzzardo) [inline README.adoc]: Add a comma after "output"

**Review by Buzzardo — CHANGES_REQUESTED** [review]

File: README.adoc
Comment (robertmcnees) [reply README.adoc]: I made the change. No objections having 2 lists would increase the clarity. I originally opted for a single list, as 90% of the instructions are the same and I don't think many users will follow the start.spring.io path anyhow.
I can see an argument either way, so I made the change. If it looks too bloated we can always revert.

File: README.adoc
Comment (robertmcnees) [reply README.adoc:124]: The Endpoint is called "Service A". I'm assuming the header is OK in general, as "Service B Endpoint" got no comments. I'll leave this as it is.

**Comment by codsaf** [comment]
### **It is so a nice work!**

**Comment by codsaf** [comment]
Can you tell me how the servicea and serviceb know the eureka server's ip address and port?

File: README.adoc
Comment (OlgaMaciaszek) [inline README.adoc]: Since you've rewritten the whole thing, it might also make sense to use `.yml` instead of properties in one or both services.

File: README.adoc
Comment (OlgaMaciaszek) [inline README.adoc]: We usually refer to it as `serviceId` ratherthan application name while discussing service discovery.

File: README.adoc
Comment (OlgaMaciaszek) [inline README.adoc]: There seems to be some formatting issue here.

File: serviceb/src/main/java/com/example/serviceb/controller/ServiceBRestController.java
Comment (OlgaMaciaszek) [inline serviceb/src/main/java/com/example/serviceb/controller/ServiceBRestController.java]: both fields could be `private final`

File: serviceb/src/main/java/com/example/serviceb/controller/ServiceBRestController.java
Comment (OlgaMaciaszek) [inline serviceb/src/main/java/com/example/serviceb/controller/ServiceBRestController.java:22]: Might make sense to verify if we have found at least one instance before running the call.

**Review by OlgaMaciaszek — APPROVED** [review]
Thanks for updating it @robertmcnees. Have added comments about some minor things, but in general looks good. I'm happy with the removal of `initial` and `complete`, however, since we're probably dealing with very inexperienced users here, I do see the benefit of having two build system versions, but that's probably more of the guides editing decision in general, so yours :) (Btw., the Cloud team usually prefers to work with Maven, but I understand that Gradle is in keeping with the starter default and that's why you've chosen it?, anyhow also more a general guides editing decision, so up to you).

File: serviceb/src/main/java/com/example/serviceb/controller/ServiceBRestController.java
Comment (robertmcnees) [reply serviceb/src/main/java/com/example/serviceb/controller/ServiceBRestController.java:22]: I think that's good defensive coding and something users should do in any example that isn't 'hello world' complexity. I'm not sure if it belongs in the guide.

If I were to expand on the code, I may do:
```
List<ServiceInstance> serviceInstances = discoveryClient.getInstances("servicea");
if(serviceInstances == null || serviceInstances.size() == 0) {
  return "no service instances found";
}
ServiceInstance serviceInstance = serviceInstances.get(0);
String serviceAResponse = ...
```

It feels like there are a lot of ways to do the check. I try for simple and verbose in the guides, so I'd turn 1 line of code into 5 to do the check. Before this edit, I haven't defined an error or fallback value if we have a problem. I could have the method throw an exception, but that adds code to the guide that isn't related to the topic and I don't explain in the README.

In my opinion I'd leave out the defensive check. A lot of our guides opt for the simplest route and leave out defensive coding practices. Current behavior if `servicea` is not available is:
```
 curl http://localhost:8082/helloEureka
{"timestamp":"2024-06-19T11:20:25.727+00:00","status":500,"error":"Internal Server Error","path":"/helloEureka"}
```
And the logs can get you there pretty quick:
```
2024-06-19T07:20:25.714-04:00 ERROR 1649 --- [serviceb] [nio-8082-exec-1] o.a.c.c.C.[.[.[/].[dispatcherServlet]    : Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed: java.lang.IndexOutOfBoundsException: Index: 0] with root cause

java.lang.IndexOutOfBoundsException: Index: 0
        at java.base/java.util.Collections$EmptyList.get(Collections.java:4586) ~[na:na]
        at com.example.serviceb.controller.ServiceBRestController.helloWorld(ServiceBRestController.java:22) ~[main/:na]
```

I'll defer to your opinion as the final word. Do you think making the check is worth the extra code? Do you have a preference on how to make the check?

**Comment by robertmcnees** [comment]
@OlgaMaciaszek I made all of the changes you requested, with the exception of the defensive coding check I'd like to get your thoughts on.

For the Gradle/Maven discussion, that was a miss on my part. I usually try to include build files in the code for both Gradle and Maven. I like to only have a single set of commands reference in the `README`, and I should default to the preference of the underlying project. In this case, I should be using Maven since that is what Spring Cloud Netflix uses.

I added the Maven wrapper and `pom.xml` to all 3 projects. I also changed the `README` so that we use Maven `./mvnw spring-boot:run` instead of Gradle `./gradlew bootRun`.

File: serviceb/src/main/java/com/example/serviceb/controller/ServiceBRestController.java
Comment (OlgaMaciaszek) [reply serviceb/src/main/java/com/example/serviceb/controller/ServiceBRestController.java:22]: Ok, we can leave it as is.

**Comment by OlgaMaciaszek** [comment]
@robertmcnees there still seem to be some gradle files present in the project. Is this on purpose?

**Comment by robertmcnees** [comment]
@OlgaMaciaszek Yes, the inclusion of both gradle and maven is on purpose. I added both build systems here for consistency to our other guides which also include both Maven and Gradle build systems. The README should feature Maven commands, according to your preference.

I believe all of our guides originated with advice from this [wiki](https://github.com/spring-guides/getting-started-guides/wiki) about cloning [this project](https://github.com/spring-guides/getting-started-guides). Several years ago the decision was made to have both build systems available in the guides.

If you feel strongly we should remove Gradle completely from this guide I can go that route. Personally, I'd prefer consistency if it doesn't add complexity for the learner. But I did break tradition here by removing the `complete` and `initial` folders because I felt that repeating all three applications in two places would be too confusing. If there is a similar argument to removing Gradle, then no problem for me to remove it.

From a sustainability perspective, I have GitHub Actions and Dependabot configured so I can manage both build systems simultaneously.

What do you think?

**Comment by OlgaMaciaszek** [comment]
Yes, it's good. May be helpful for the users to have both build systems.

**Pull request merged**
